### ISSUE TITLE
Release for ubuntu linux

### ISSUE THREAD

**Issue by badushaebrahim** [Sangwan5688/BlackHole#311]
how to build app for linux/ubuntu/pop os

**Comment by dalpat** [comment]
I tried with `flutter pub get` and `flutter run -d linux` still no luck.

`flutter run -d chrome` this opened a blank page in chrome

on ubuntu 22.04 LTS.

Attached the error SS of `flutter run -d linux`
![image](https://user-images.githubusercontent.com/6994666/169961380-1daa5515-6c0e-493d-8952-84d3832f78cf.png)

**Comment by badushaebrahim** [comment]
yep same error

**Comment by badushaebrahim** [comment]
![shr](https://user-images.githubusercontent.com/63869474/170875116-5a8c549d-6e4f-41a5-975c-d7e1069156a8.jpeg)

**Comment by badushaebrahim** [comment]
mines a little different but your can be fixed by installing flutter from flutter git and not snap store

**Comment by Sangwan5688** [comment]
@badushaebrahim, you can change the directory name from "BlackHole" to "blackhole". It will work fine then. As for other issues, I haven't tried building the Linux version, nor am I sure if all the packages and stuff will work on Linux.

**Comment by badushaebrahim** [comment]
I changed and now its cmake error

**Comment by badushaebrahim** [comment]
@dalpat  are you using flutter from snap store or github

**Comment by dalpat** [comment]
I used the tarball from the official docs [[linux_tarball]](https://docs.flutter.dev/get-started/install/linux)

because snap apps are known for their delayed startup issues, I avoid anything snap. so I installed it the manual way as described by official docs.

**Comment by dalpat** [comment]
The problem may be with the libwinmedia version.

line 36 says 0.0.4 but the error is 0.0.3

![image](https://user-images.githubusercontent.com/6994666/171981431-e78dacf6-8402-4e8a-b085-33159aa1b09b.png)

**Comment by badushaebrahim** [comment]
@dalpat  CAN You explain

**Comment by dalpat** [comment]
I am no good with `flutter` so I may be wrong.

As we can see from the above SS, the error is `libwinmedia - 0.0.3` but in `pubspec.yml` on line 36 `just_audio_libwinmedia: ^0.0.4` is required.
Maybe `0.0.4` provide some functionality upon which `blackhole` is dependent, which may not be present in `0.0.3`.

Currently, I'm on `ubuntu 22.04 LTS` and I don't know how to get `0.0.4` so that I can build again to test if it works with  `0.0.4`.

**Comment by Sangwan5688** [comment]
See, the thing is blackhole uses `just_audio` along with `audio_service` to
play audio on Android and iOS. But these packages don't support windows or
Linux. Whereas there's a package called `libwinmedia` which is used in
windows and Linux to play audio. Now, that's where `libwinmedia_just_audio`
comes in. It is a package which uses libwinmedia for windows and Linux and
combines it with `just_audio` and `audio_service` so that the code which
you have written for Android and iOS can be used for windows and Linux too.
But that package is still in development. And maybe they even dropped it.
Whatever the case, it's not complete and has many bugs which causes these
issues on windows and Linux.

**Comment by badushaebrahim** [comment]
@dal man we can try "sudo apt-get install libwebkit2gtk-4.0-dev" run this a nd try and change 0.0.4 version and then it might run
at locally

**Comment by dalpat** [comment]
> @dal man we can try "sudo apt-get install libwebkit2gtk-4.0-dev" run this a nd try and change 0.0.4 version and then it might run at locally

yeah, I just tried from [https://github.com/bdlukaa/just_audio_libwinmedia#linux](https://github.com/bdlukaa/just_audio_libwinmedia#linux)

`sudo apt-get install libwebkit2gtk-4.0-dev`

and the error gone, but gave me permission error to `/usr/local`

so I gave `a+w` to `/usr/local`

then it gave permission error on `/usr/local/lib`, I gave `a+w` to `/usr/local/lib`

and now I'm stuck with this new error

![image](https://user-images.githubusercontent.com/6994666/174003605-5def50f8-8b2d-4190-97c1-1d0016a382fd.png)


I changed the version to 0.0.3 in `pubspec.yml` see line 36 in below SS

The terminal prompt wait for a while at `Building linux application...` message and then gives this error

![image](https://user-images.githubusercontent.com/6994666/174004280-a14b0916-7a7b-47de-b898-03753e711e0c.png)





Does anyone have any Idea how to proceed further from here?

**Comment by topcheese** [comment]
I just found this project today, perhaps we can integrate it for this issue? One of us in the community, although I'm pretty sure my skills are not up to it yet.
https://pub.dev/packages/audioplayers

**Comment by badushabjelt** [comment]
@Sangwan5688  can we use the updated versiion of teh package

**Comment by badushaebrahim** [comment]
@Sangwan5688

**Comment by Sangwan5688** [comment]
Umm, I picked up just_audio and audio_service after a lot of research. Some packages don't play in background, some don't support features like adding to queue and stuff. And the whole logic is based on just_audio and audio_service. So it will be a lot of work to switch to any other package.

**Comment by Sangwan5688** [comment]
Instead if someone wants they can change the code just for windows and linux while keeping the current one intact for android and iOS. We can slowly add all current features for desktop using that package. And once all those features are present we can move the android and iOS side logic too.

**Comment by badushaebrahim** [comment]
ok

**Comment by RaptaG** [comment]
Is Linux support still planned?

**Comment by Ranger-NF** [comment]
> Instead if someone wants they can change the code just for windows and linux while keeping the current one intact for android and iOS. We can slowly add all current features for desktop using that package. And once all those features are present we can move the android and iOS side logic too.

Im trying to develop Blackhole for linux using [audioplayers](https://pub.dev/packages/audioplayers). Im sure this is gonna take a long time, but I will update if I get any progress done :)

**Comment by AnAverageBeing** [comment]
@Ranger-NF any updates

**Comment by Ranger-NF** [comment]
Not really, I have been looking at the source to understand the structure of the app while trying out the features of `audioplayers`. Its time consuming and currently Im  mainly focusing  on my main project ;-;

**Comment by Ranger-NF** [comment]
@AnAverageBeing I have tried changing code on the project after my last comment. And now song playback works (You can play a song at once), Its just that now I have to try and somehow get the current queue system to work with the new package

Should I upload the changes to a fork repo in my account for anyone interested in trying it out ?

**Comment by AnAverageBeing** [comment]
@Ranger-NF  yes why not do it i will try out

**Comment by Ranger-NF** [comment]
@AnAverageBeing Ok so, I have pushed the code to a repo: https://github.com/Ranger-NF/Blackhole-Linux 
I will try and do some work on it in the coming days :)

**Comment by Ranger-NF** [comment]
Ok so, I spent some time trying to replace just-audio with audioplayers but failed. I know this is unfortunate, but yeah... Im gonna call it a quit from my side coz of my inexperience in flutter and some parts of the code are not  documented (with comments) which  makes it even harder

NOTE: If anyone else is trying to take on this challenge, please let me know...

**Comment by BrightDV** [comment]
> Ok so, I spent some time trying to replace just-audio with audioplayers but failed. I know this is unfortunate, but yeah... Im gonna call it a quit from my side coz of my inexperience in flutter and some parts of the code are not documented (with comments) which makes it even harder
> 
> NOTE: If anyone else is trying to take on this challenge, please let me know...

I will try to see how you started implementing it and maybe try to do it. If I achieve to do some progress, I will open a PR in your repo!

**Comment by BrightDV** [comment]
So now the playback **works** (using the mpv player and just_audio). See https://github.com/Ranger-NF/Blackhole-Linux/pull/2 for the fix.

**Comment by Ranger-NF** [comment]
> So now the playback **works** (using the mpv player and just_audio). See [Ranger-NF/Blackhole-Linux#2](https://github.com/Ranger-NF/Blackhole-Linux/pull/2) for the fix, if you want to test.

The PR made by @BrightDV has been merged. And like BrightDV mentioned earlier, playback works and the app seems to handle most use cases. But its no where near perfect.

You can clone the [repo](https://github.com/Ranger-NF/Blackhole-Linux) to try it out for yourself.
